File: common/src/main/java/io/github/mqzn/commands/utilities/TimeParser.java
package io.github.mqzn.commands.utilities;

import java.util.concurrent.TimeUnit;

/**
 * The main class that is used to parse time strings inputs
 * The algorithm is fully written by me (aka Mqzen)
 * Make sure to know just when to use it properly :D !
 *
 * @author Mqzen
 */
public class TimeParser {
	private long days;
	private long hours;
	private long minutes;
	private long seconds;
	
	private TimeParser(long millis) {
		this.days = millis / 86400000L;
		this.hours = (millis / 3600000L) % 24L;
		this.minutes = (millis / 60000L) % 60L;
		this.seconds = (millis / 1000L) % 60L;
	}
	
	private TimeParser(String timePeriod) {
		char[] chars = timePeriod.toCharArray();
		int i = 0;
		while (i < timePeriod.length()) {
			if (Character.isDigit(chars[i])) {
				StringBuilder digitToCollect = new StringBuilder();
				int start = i;
				while (Character.isDigit(chars[start])) {
					digitToCollect.append(chars[start]);
					start++;
				}
				
				char unit = chars[start];
				int digit = Integer.parseInt(digitToCollect.toString());
				switch (unit) {
					case 'd', 'D' -> this.days += digit;
					case 'h', 'H' -> this.hours += digit;
					case 'm', 'M' -> this.minutes += digit;
					case 's', 'S' -> this.seconds += digit;
				}
				i = start;
			}
			i++;
		}
	}
	
	public static TimeParser parse(String timePeriod) {
		return new TimeParser(timePeriod);
	}
	
	public static TimeParser parse(long millis) {
		return new TimeParser(millis);
	}
	
	public Pair<Long, TimeUnit> highestLogicalUnitValue() {
		if (this.days != 0L) {
			return Pair.of(this.days, TimeUnit.DAYS);
		}
		if (this.hours != 0L) {
			return Pair.of(this.hours, TimeUnit.HOURS);
		}
		if (this.minutes != 0L) {
			return Pair.of(this.minutes, TimeUnit.MINUTES);
		}
		return Pair.of(this.seconds, TimeUnit.SECONDS);
	}
}

File: common/src/main/java/io/github/mqzn/commands/utilities/TimeParser.kt
package io.github.mqzn.commands.utilities

import lombok.Getter
import java.util.concurrent.TimeUnit

/**
 * The main class that is used to parse time strings inputs
 * The algorithm is fully written by me (aka Mqzen)
 * Make sure to know just when to use it properly :D !
 *
 * @author Mqzen
 */
class TimeParser {
    @Getter
    private var days: Long = 0

    @Getter
    private var hours: Long = 0

    @Getter
    private var minutes: Long = 0

    @Getter
    private var seconds: Long = 0

    private constructor(millis: Long) {
        days = millis / 86400000L
        hours = millis / 3600000L % 24L
        minutes = millis / 60000L % 60L
        seconds = millis / 1000L % 60L
    }

    private constructor(timePeriod: String) {
        val chars = timePeriod.toCharArray()
        var i = 0
        while (i < timePeriod.length) {
            if (Character.isDigit(chars[i])) {
                val digitToCollect = StringBuilder()
                var start = i
                while (Character.isDigit(chars[start])) {
                    digitToCollect.append(chars[start])
                    start++
                }

                //the current index is the end of the digit to collect
                //so the current index is that of a unit char
                val unit = chars[start]
                val digit = digitToCollect.toString().toInt()
                when (unit) {
                    'd', 'D' -> days += digit.toLong()
                    'h', 'H' -> hours += digit.toLong()
                    'm', 'M' -> minutes += digit.toLong()
                    's', 'S' -> seconds += digit.toLong()
                }
                i = start
            }
            i++
        }
    }

    fun highestLogicalUnitValue(): Pair<Long, TimeUnit> {
        if (days != 0L) return Pair.of(days, TimeUnit.DAYS)
        if (hours != 0L) return Pair.of(hours, TimeUnit.HOURS)
        return if (minutes != 0L) Pair.of(minutes, TimeUnit.MINUTES) else Pair.of(seconds, TimeUnit.SECONDS)
    }

    companion object {
        fun parse(timePeriod: String): TimeParser {
            return TimeParser(timePeriod)
        }

        @JvmStatic
        fun parse(millis: Long): TimeParser {
            return TimeParser(millis)
        }
    }
}